src/dbnode/storage/namespace.go

```diff
@@ -1641,134 +1641,60 @@ func (n *dbNamespace) aggregateTiles(
-}
+		sourceBlockVolumes := make([]shardBlockVolume, 0, len(sourceBlockStarts))
+		for _, sourceBlockStart := range sourceBlockStarts {
+			latestVolume, err := sourceShard.LatestVolume(sourceBlockStart)
+			if err != nil {
+				n.log.Error("error getting shards latest volume",
+					zap.Error(err), zap.Uint32("shard", sourceShard.ID()), zap.Time("blockStart", sourceBlockStart))
+				return 0, err
+			}
+			sourceBlockVolumes = append(sourceBlockVolumes, shardBlockVolume{sourceBlockStart, latestVolume})
+		}
 
-func (n *dbNamespace) coldFlushSingleShard(
-	nsCtx namespace.Context,
-	shard databaseShard,
-	pm persist.Manager,
-	resources coldFlushReuseableResources,
-	multiErr xerrors.MultiError,
-) xerrors.MultiError {
-	// NB(rartoul): This value can be used for emitting metrics, but should not be used
-	// for business logic.
-	callStart := n.nowFn()
+		shardprocessedTileCount, err := targetShard.AggregateTiles(ctx, sourceNs.ID(), sourceShard.ID(), blockReaders, sourceBlockVolumes, opts, nsCtx.Schema)
 
-	// NB(bodu): The in-mem index will lag behind the TSDB in terms of new series writes. For a period of
-	// time between when we rotate out the active cold mutable index segments (happens here) and when
-	// we actually cold flush the data to disk we will be making writes to the newly active mutable seg.
-	// This means that some series can live doubly in-mem and loaded from disk until the next cold flush
-	// where they will be evicted from the in-mem index.
-	var (
-		onColdFlushDone OnColdFlushDone
-		err             error
-	)
-	if n.reverseIndex != nil {
-		onColdFlushDone, err = n.reverseIndex.ColdFlush([]databaseShard{shard})
+		processedTileCount += shardprocessedTileCount
 		if err != nil {
-			n.metrics.aggregateTiles.ReportError(n.nowFn().Sub(callStart))
-			return multiErr.Add(
-				fmt.Errorf("error preparing to coldflush a reverse index for shard %d: %v",
-					shard.ID(),
-					err))
+			detailedErr := fmt.Errorf("shard %d aggregation failed: %v", targetShard.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+			continue
 		}
 	}
 
-	onColdFlushNs, err := n.opts.OnColdFlush().ColdFlushNamespace(n)
-	if err != nil {
-		n.metrics.aggregateTiles.ReportError(n.nowFn().Sub(callStart))
-		return multiErr.Add(
-			fmt.Errorf("error preparing to coldflush a namespace for shard %d: %v",
-				shard.ID(),
-				err))
-	}
-
-	flushPersist, err := pm.StartFlushPersist()
-	if err != nil {
-		n.metrics.aggregateTiles.ReportError(n.nowFn().Sub(callStart))
-		return multiErr.Add(
-			fmt.Errorf("error starting flush persist for shard %d: %v",
-				shard.ID(),
-				err))
-	}
-
-	localErrors := xerrors.NewMultiError()
-	shardColdFlush, err := shard.ColdFlush(flushPersist, resources, nsCtx, onColdFlushNs)
+	err := multiErr.FinalError()
 	if err != nil {
-		detailedErr := fmt.Errorf("shard %d failed to compact: %v", shard.ID(), err)
-		localErrors = localErrors.Add(detailedErr)
+		n.log.Error("errors when aggregating large tiles", zap.Error(err))
```

Comment:
This is going to duplicate the errors already logged within the shard. Can we do the per shard errors in one place?

Reply:
👍